<a href="https://colab.research.google.com/github/RachelDojo/Master-SQL-Development/blob/master/Python_for_DataScience_ch1%262.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

""" linkedin course
Path: https://www.linkedin.com/learning/paths/advance-your-python-skills-for-data-science

Python for Data Science Essential Training Part 1:
https://www.linkedin.com/learning/python-for-data-science-essential-training-part-1/data-science-life-hacks?contextUrn=urn%3Ali%3AlyndaLearningPath%3A5b61ea25498e580437e51859


## Python for Data Science Essential training
# Chapter 1. Introduction to the data professions
Four functions work together to generate business value from data:
/Data scrience
/Data engineering
/Data analytics
/Business intelligence

You need to choose one direction to start, don't try to grab all of them. Why?
- You are the one that gets to decide where you best fit within the data professions - be deliberate
- Less confusion, more results

Data science is the systematic study of the structure and behavior of data in order to quantifiably 可量化地 understand past and current occurrences, as well as predict the future behavior of that data

# Data scientist:
- degree in a quantitative field
- programmers: R, Python etc.
- Curious and tenacious 顽强 personalities
- previous experience: BI reporting, data analysis, and data-driven decision-making
- Interested in why over how
- Spends hours analyzing complex questions
- Data scientists derive insights from data, including big data (data from data engineered systems, many different structures, velocities and varieties in a more innovative system); while BI people and statisticien people are pulling data from relational databases.  
- Uncover correlations and causations in business data to support business decision-making
- generate predictions from data and communicate predictions through data visualization
- STEM degree with advanced math:calculation, differencial, linear algebra
- Coding: Python, R, SQL, Spark, machine learning

# Data Engineering
Data engineering is the design, construction, and maintenance of data systems
- programmers
- good in math, but don't use statistics much
- previous experience: database architech, ETL developer, database developer, security architect etc.
- Prefer to degin and build IT systems rather than to analyze data
- more interested in how over why
- data engineers design systems to collect, handle and store big datasets
- build modular, scalable platforms for data processing
- design, build and maintain systems that store and moving big data
- computer science, software engineering
- coding: java, C++, Python

# Data analytics (analytic specials or analytics expert)
Data analytics are data products that describe data and how it behaves. These data products are generated from data analysis and visualization processes.
- Use applications to analyze data
- Solid in basic math
- Understan the inner workings of a business very well
- they interface between the advanced and technical inner workings of the data scientists and engineers
- the liaison 联络 between the more complex data roles and the business requirements
- they enjoying using data insights to improve business.
- Not into deep analysis
- Generates predictive, prescriptive and descriptive data insights
- Uncover correlations and causations in business data to support business decision-making
- Deploys analytics technologies: analytics software applications




## Why use Python for data analysis?

- Coding languages: Python, R, Go, Julia
- Python is a high-level, interpreted coding language that is useful for a very wide variety of applications; is one of the official programming languages of Google

- Benefit of Using Python
  - Easy to learn, human readable
  - Extensive array of well-supported data science libraries
  - Biggest user base of all data science languages
  - Useful in data engineering and data science

- Main Python Libraries for Data Science
  - Advanced Data Analysis: NumPy, SciPy, pandas
  - Data Visualization: Matplotlib, Seaborn
  - Machine Learning: scikit-learn, TensorFlow, Keras
  - Deep learning: TensorFlow, Keras



## Chapter 2. Data Preparation Basics

# 2.1 Filtering and selecting

# Introducing the pandas Library
- Why pandas: fast data cleaning, preparation and analysis; easy to use for data visualization and machine learning
- What is pandas: built on top of NumPy, makes it easy to work with arrays and matrics
- Indexing in pandas
  -An index is a list of integers or labels you use to uniquely identify rows or columns
  - in this course: square-brackets [...] or .loc[ ] indexer
  - DataFrame Objects: like a spreaksheet in Excel, made of a set of series objects and are indexable
  - Series objects: a row or column that is indexed

***NumPy is the fundamental package for scientific computing in Python. It is a Python library that provides a multidimensional array object, various derived objects (such as masked arrays and matrices), and an assortment of routines for fast operations on arrays, including mathematical, logical, shape manipulation, sorting, selecting, I/O, discrete Fourier transforms, basic linear algebra, basic statistical operations, random simulation and much more.

***Pandas is a Python library used for working with data sets. It has functions for analyzing, cleaning, exploring, and manipulating data.
The name "Pandas" has a reference to both "Panel 面板 Data", and "Python Data Analysis" and was created by Wes McKinney in 2008.

In [ ]:
import numpy as np
import pandas as pd

from pandas import Series, DataFrame

## to use Shift + enter to import the libraries to the iPython environment

In [ ]:
## selecting and retrieving 取回 data
# create series index
series_obj = Series(np.arange(8),index = ["row1", "row2", "row3", "row4", "row5", "row6", "row7", "row8"])

# call the index
series_obj



row1    0
row2    1
row3    2
row4    3
row5    4
row6    5
row7    6
row8    7
dtype: int64

In [ ]:
#using [lable index ] to select the lable index

series_obj["row7"]



6

In [ ]:
#using [ integer] to select and retrieve all records with this integer index
series_obj[[0,7]]



row1    0
row8    7
dtype: int64

In [ ]:
# To create DataFrame object with 36 random numbers
np.random.seed(25)
DF_obj = DataFrame(np.random.rand(36).reshape((6,6)),
                   index=["row1", "row2", "row3", "row4", "row5", "row6"],
                   columns=["column1", "column2", "column3", "column4", "column5", "column6"])
DF_obj


column1   column2   column3   column4   column5   column6
row1  0.870124  0.582277  0.278839  0.185911  0.411100  0.117376
row2  0.684969  0.437611  0.556229  0.367080  0.402366  0.113041
row3  0.447031  0.585445  0.161985  0.520719  0.326051  0.699186
row4  0.366395  0.836375  0.481343  0.516502  0.383048  0.997541
row5  0.514244  0.559053  0.034450  0.719930  0.421004  0.436935
row6  0.281701  0.900274  0.669612  0.456069  0.289804  0.525819

In [ ]:
# Select rows or columns from a data frame
DF_obj.loc[["row2","row5"], ["column5", "column2"]]

column5   column2
row2  0.402366  0.437611
row5  0.421004  0.559053

##Data slicing 切片
To select and return a slice of several values from a data set. Slicing uses indes values with [ : ]

With slicing you pass in two index values that are separated by a colon. The index value on the left side of the colon should be the first value selected; on the right of the colon the last value selected. When execute the code, the indexer simply finds the first and the last record and returns in between them.

In [ ]:
series_obj["row3":"row7"]

row3    2
row4    3
row5    4
row6    5
row7    6
dtype: int64

##Comparing with scalars
Scalar value is just a single numerical value
Comparison operators: > <
To compare each value with a scalar value then get the result of boelean values of True or false

In [ ]:
DF_obj < 0.2


column1  column2  column3  column4  column5  column6
row1    False    False    False     True    False     True
row2    False    False    False    False    False     True
row3    False    False     True    False    False    False
row4    False    False    False    False    False    False
row5    False    False     True    False    False    False
row6    False    False    False    False    False    False

#Filtering with scalars
To only return records that satisfy the comparison expression

In [ ]:
series_obj[series_obj > 6]

row8    7
dtype: int64

## Setting values with scalars

Setting is where you select all records associated with a specified label index and then set those values equal to a scalar

In [ ]:
 series_obj["row1"] = 8
 series_obj

row1    8
row2    1
row3    2
row4    3
row5    4
row6    5
row7    6
row8    7
dtype: int64

## 2.2 Treating missing Values in Python
- By default, missing values are represented with NaN:"Not a Number"
- Warning: If your dataset has 0s, 99s or 999s, be sure to either drop or approximate them as you would with missing values
- Way 1: take the average missing values in order to generate an approximation

In [ ]:
import numpy as np
import pandas as pd

from pandas import Series, DataFrame

In [ ]:
##Figuring out what data is missing

missing = np.nan
series_obj = Series(['row1', 'row2', missing, 'row4', 'row5', 'row6', missing, 'row8'])
series_obj


0    row1
1    row2
2     NaN
3    row4
4    row5
5    row6
6     NaN
7    row8
dtype: object

In [ ]:
#isnull method: runs a boolean value that describes in true or false whether an element in a Pandas object is a null value (null value is missing value)

series_obj.isnull()

0    False
1    False
2     True
3    False
4    False
5    False
6     True
7    False
dtype: bool

In [ ]:
#How to fill in for missing values
## create a data frame
np.random.seed(25)
DF_obj = DataFrame(np.random.rand(36).reshape(6,6))
DF_obj

0         1         2         3         4         5
0  0.870124  0.582277  0.278839  0.185911  0.411100  0.117376
1  0.684969  0.437611  0.556229  0.367080  0.402366  0.113041
2  0.447031  0.585445  0.161985  0.520719  0.326051  0.699186
3  0.366395  0.836375  0.481343  0.516502  0.383048  0.997541
4  0.514244  0.559053  0.034450  0.719930  0.421004  0.436935
5  0.281701  0.900274  0.669612  0.456069  0.289804  0.525819

In [ ]:
##with .loc method to select rows and columns and then set certain values equal to missing (in the ['row value', 'column value])

DF_obj.loc[3:5, 0] = missing
DF_obj.loc[1:4, 5] = missing
DF_obj


0         1         2         3         4         5
0  0.870124  0.582277  0.278839  0.185911  0.411100  0.117376
1  0.684969  0.437611  0.556229  0.367080  0.402366       NaN
2  0.447031  0.585445  0.161985  0.520719  0.326051       NaN
3       NaN  0.836375  0.481343  0.516502  0.383048       NaN
4       NaN  0.559053  0.034450  0.719930  0.421004       NaN
5       NaN  0.900274  0.669612  0.456069  0.289804  0.525819

In [ ]:
## With .fillna() method to fill in the missing values with a numeric value that you specify
# in this case fill in missing value with 0

filled_DF = DF_obj.fillna(0)
filled_DF

0         1         2         3         4         5
0  0.870124  0.582277  0.278839  0.185911  0.411100  0.117376
1  0.684969  0.437611  0.556229  0.367080  0.402366  0.000000
2  0.447031  0.585445  0.161985  0.520719  0.326051  0.000000
3  0.000000  0.836375  0.481343  0.516502  0.383048  0.000000
4  0.000000  0.559053  0.034450  0.719930  0.421004  0.000000
5  0.000000  0.900274  0.669612  0.456069  0.289804  0.525819

In [ ]:
#fill in the missing values from each column series as designated by the dictionary key with its own value as specified by the corresponding dictionary value{}

filled_DF = DF_obj.fillna({0:0.1, 5:1.25})
filled_DF


0         1         2         3         4         5
0  0.870124  0.582277  0.278839  0.185911  0.411100  0.117376
1  0.684969  0.437611  0.556229  0.367080  0.402366  1.250000
2  0.447031  0.585445  0.161985  0.520719  0.326051  1.250000
3  0.100000  0.836375  0.481343  0.516502  0.383048  1.250000
4  0.100000  0.559053  0.034450  0.719930  0.421004  1.250000
5  0.100000  0.900274  0.669612  0.456069  0.289804  0.525819

How this could be useful?
- predictive application needs 4 variables as input, one variable has missing values. To fill in an approximation in order to make the predictive app work
- pass in the method 'ffill' argument  

In [ ]:
# 'ffill' method: fill in missing values with values from the last non-null element in the column series

fill_DF = DF_obj.fillna(method='ffill')
fill_DF



0         1         2         3         4         5
0  0.870124  0.582277  0.278839  0.185911  0.411100  0.117376
1  0.684969  0.437611  0.556229  0.367080  0.402366  0.117376
2  0.447031  0.585445  0.161985  0.520719  0.326051  0.117376
3  0.447031  0.836375  0.481343  0.516502  0.383048  0.117376
4  0.447031  0.559053  0.034450  0.719930  0.421004  0.117376
5  0.447031  0.900274  0.669612  0.456069  0.289804  0.525819

##Counting missing values
Useful:
- creating a statistic summary of your dataset - count missing values to figure out which varaibles are most problematic.

In [ ]:
# again create a dataset
np.random.seed(25)
DF_obj = DataFrame(np.random.rand(36).reshape(6,6))
DF_obj

0         1         2         3         4         5
0  0.870124  0.582277  0.278839  0.185911  0.411100  0.117376
1  0.684969  0.437611  0.556229  0.367080  0.402366  0.113041
2  0.447031  0.585445  0.161985  0.520719  0.326051  0.699186
3  0.366395  0.836375  0.481343  0.516502  0.383048  0.997541
4  0.514244  0.559053  0.034450  0.719930  0.421004  0.436935
5  0.281701  0.900274  0.669612  0.456069  0.289804  0.525819

In [ ]:
#generate missing values in the dataset
DF_obj.loc[3:5, 0] = missing
DF_obj.loc[1:4, 5] = missing
DF_obj

0         1         2         3         4         5
0  0.870124  0.582277  0.278839  0.185911  0.411100  0.117376
1  0.684969  0.437611  0.556229  0.367080  0.402366       NaN
2  0.447031  0.585445  0.161985  0.520719  0.326051       NaN
3       NaN  0.836375  0.481343  0.516502  0.383048       NaN
4       NaN  0.559053  0.034450  0.719930  0.421004       NaN
5       NaN  0.900274  0.669612  0.456069  0.289804  0.525819

In [ ]:
# generate a count of missing values with .sum() method

DF_obj.isnull().sum()



0    3
1    0
2    0
3    0
4    0
5    4
dtype: int64

In [ ]:
## How to filter/drop out all rows missing values with dropna method

DF_no_NaN = DF_obj.dropna()
DF_no_NaN


0         1         2         3       4         5
0  0.870124  0.582277  0.278839  0.185911  0.4111  0.117376

In [ ]:
# to drop out all columns with missing values

DF_no_NaN = DF_obj.dropna(axis=1)
DF_no_NaN

1         2         3         4
0  0.582277  0.278839  0.185911  0.411100
1  0.437611  0.556229  0.367080  0.402366
2  0.585445  0.161985  0.520719  0.326051
3  0.836375  0.481343  0.516502  0.383048
4  0.559053  0.034450  0.719930  0.421004
5  0.900274  0.669612  0.456069  0.289804

##2.3 Removing duplicates
- To maintain accurate, consistent datasets
- To avoid producing erroneous or misleading statistics


In [ ]:
# To generate a dataset with dictionary{ }
DF_obj = DataFrame({'column 1': [1, 1, 2, 2, 3, 3, 3],
                  'column 2': ['a', 'a', 'b', 'b', 'c', 'c', 'c'],
                  'column 3': ['A', 'A', 'B', 'B', 'C', 'C', 'C']})
DF_obj

column 1 column 2 column 3
0         1        a        A
1         1        a        A
2         2        b        B
3         2        b        B
4         3        c        C
5         3        c        C
6         3        c        C

In [ ]:
# to find out the duplicates rows with .duplicated() method
# The .duplicated() method searches each row in the DataFrame, and returns a True or False value to
#indicate whether it is a duplicate of another row found earlier in the DataFrame.

DF_obj.duplicated()

0    False
1     True
2    False
3     True
4    False
5     True
6     True
dtype: bool

In [ ]:
# to drop the duplicated rows above
DF_obj.drop_duplicates()

column 1 column 2 column 3
0         1        a        A
2         2        b        B
4         3        c        C

In [ ]:
# to drop cuplicated columns
# generate dataset
DF_obj = DataFrame({'column 1': [1, 1, 2, 2, 3, 3, 3],
                  'column 2': ['a', 'a', 'b', 'b', 'c', 'c', 'c'],
                  'column 3': ['A', 'A', 'B', 'B', 'C', 'D', 'C']})
DF_obj



column 1 column 2 column 3
0         1        a        A
1         1        a        A
2         2        b        B
3         2        b        B
4         3        c        C
5         3        c        D
6         3        c        C

In [ ]:
# to drop the duplicates in column 3

DF_obj.drop_duplicates(['column 3'])

column 1 column 2 column 3
0         1        a        A
2         2        b        B
4         3        c        C
5         3        c        D

## Concatenate and transform data

- Concatenation: combining data from seperate sources
- Transformation is converting and reforming data to the format that's neccessary

In [ ]:
import numpy as np
import pandas as pd

from pandas import Series, DataFrame

#generate dataset

DF_obj = pd.DataFrame(np.arange(36).reshape(6,6))
DF_obj

0   1   2   3   4   5
0   0   1   2   3   4   5
1   6   7   8   9  10  11
2  12  13  14  15  16  17
3  18  19  20  21  22  23
4  24  25  26  27  28  29
5  30  31  32  33  34  35

In [ ]:
# to concatenate this dataset

DF_obj_2 = pd.DataFrame(np.arange(15).reshape(5,3))
DF_obj_2

0   1   2
0   0   1   2
1   3   4   5
2   6   7   8
3   9  10  11
4  12  13  14

#Concatenating data with pd.concat([left_object, right_object], axis=1)

- concat() method joins data from seperate sources into one combined data table
- joining row datas: pd.concat()
- axis=1 tells Python to concatenate the DataFrame by adding columns (in other words, joining on the row index values).

In [ ]:
pd.concat([DF_obj, DF_obj_2], axis =1)

0   1   2   3   4   5     0     1     2
0   0   1   2   3   4   5   0.0   1.0   2.0
1   6   7   8   9  10  11   3.0   4.0   5.0
2  12  13  14  15  16  17   6.0   7.0   8.0
3  18  19  20  21  22  23   9.0  10.0  11.0
4  24  25  26  27  28  29  12.0  13.0  14.0
5  30  31  32  33  34  35   NaN   NaN   NaN

In [ ]:
# to concatenate based on the column values

pd.concat([DF_obj, DF_obj_2])

0   1   2     3     4     5
0   0   1   2   3.0   4.0   5.0
1   6   7   8   9.0  10.0  11.0
2  12  13  14  15.0  16.0  17.0
3  18  19  20  21.0  22.0  23.0
4  24  25  26  27.0  28.0  29.0
5  30  31  32  33.0  34.0  35.0
0   0   1   2   NaN   NaN   NaN
1   3   4   5   NaN   NaN   NaN
2   6   7   8   NaN   NaN   NaN
3   9  10  11   NaN   NaN   NaN
4  12  13  14   NaN   NaN   NaN

## Transforming data
#### Dropping data


In [ ]:
# to drop row based indexes

DF_obj.drop([0,2])

0   1   2   3   4   5
1   6   7   8   9  10  11
3  18  19  20  21  22  23
4  24  25  26  27  28  29
5  30  31  32  33  34  35

In [ ]:
# to drop clumns with axis=1

DF_obj.drop([0,2], axis=1)

1   3   4   5
0   1   3   4   5
1   7   9  10  11
2  13  15  16  17
3  19  21  22  23
4  25  27  28  29
5  31  33  34  35

#### Adding data with .join(object1, object2) & .append()

!!!The frame.append method is deprecated (abandoned) and will be removed from pandas in a future version. Use pandas.concat instead

In [ ]:
# to create a series data to add into dataset
series_obj = Series(np.arange(6))
series_obj.name = "added_variable"
series_obj

0    0
1    1
2    2
3    3
4    4
5    5
Name: added_variable, dtype: int64

In [ ]:
# join this new variable into data sources with .join()

variable_added = DataFrame.join(DF_obj, series_obj)
variable_added



0   1   2   3   4   5  added_variable
0   0   1   2   3   4   5               0
1   6   7   8   9  10  11               1
2  12  13  14  15  16  17               2
3  18  19  20  21  22  23               3
4  24  25  26  27  28  29               4
5  30  31  32  33  34  35               5

In [ ]:
# two ways to use the .append method
# one is leave the original index in place and the other is to recreate the index

# recreate a new data table

added_datatable = variable_added.append(variable_added, ignore_index=False)
added_datatable


<ipython-input-38-14b068dda13e>:6: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  added_datatable = variable_added.append(variable_added, ignore_index=False)


0   1   2   3   4   5  added_variable
0   0   1   2   3   4   5               0
1   6   7   8   9  10  11               1
2  12  13  14  15  16  17               2
3  18  19  20  21  22  23               3
4  24  25  26  27  28  29               4
5  30  31  32  33  34  35               5
0   0   1   2   3   4   5               0
1   6   7   8   9  10  11               1
2  12  13  14  15  16  17               2
3  18  19  20  21  22  23               3
4  24  25  26  27  28  29               4
5  30  31  32  33  34  35               5

In [ ]:
# see the difference between above and this one. the index has been reset

added_datatable = variable_added.append(variable_added, ignore_index=True)
added_datatable

<ipython-input-36-b2d1bccdcfb9>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  added_datatable = variable_added.append(variable_added, ignore_index=True)


0   1   2   3   4   5  added_variable
0    0   1   2   3   4   5               0
1    6   7   8   9  10  11               1
2   12  13  14  15  16  17               2
3   18  19  20  21  22  23               3
4   24  25  26  27  28  29               4
5   30  31  32  33  34  35               5
6    0   1   2   3   4   5               0
7    6   7   8   9  10  11               1
8   12  13  14  15  16  17               2
9   18  19  20  21  22  23               3
10  24  25  26  27  28  29               4
11  30  31  32  33  34  35               5

#### Sorting data with .sort_values()
To sort rows in a DataFrame, either in ascending or descending order, call the .sort_values() method of the DataFrame, and pass in the by argument to specify the column index upon which the DataFrame should be sorted.


In [ ]:
# sort the data frame by values in column 5, are sorted descending
DF_sorted = DF_obj.sort_values(by=[5], ascending=[False])
DF_sorted

0   1   2   3   4   5
5  30  31  32  33  34  35
4  24  25  26  27  28  29
3  18  19  20  21  22  23
2  12  13  14  15  16  17
1   6   7   8   9  10  11
0   0   1   2   3   4   5

#### Grouping and aggregation

For exploring and describing dataset in its subgroups:
- to compare subsets
- to deduce reasons why subgroups differ
- to ubset data for analysis

In [ ]:
import numpy as np
import pandas as pd
from pandas import Series, DataFrame, io

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
address = pd.read_csv('mtcars.csv')

print(address)

              car_names   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  \
0             Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1   
1         Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1   
2            Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1   
3        Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0   
4     Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0   
5               Valiant  18.1    6  225.0  105  2.76  3.460  20.22   1   0   
6            Duster 360  14.3    8  360.0  245  3.21  3.570  15.84   0   0   
7             Merc 240D  24.4    4  146.7   62  3.69  3.190  20.00   1   0   
8              Merc 230  22.8    4  140.8   95  3.92  3.150  22.90   1   0   
9              Merc 280  19.2    6  167.6  123  3.92  3.440  18.30   1   0   
10            Merc 280C  17.8    6  167.6  123  3.92  3.440  18.90   1   0   
11           Merc 450SE  16.4    8  275.8  180  3.07  4.070  17.

In [ ]:
from google.colab import files
uploaded = files.upload()


Saving mtcars.csv to mtcars.csv


In [ ]:
# Grouping data by column index:  object_name.groupby('Series_name')

import pandas as pd

import io

cars = pd.read_csv(io.BytesIO(uploaded['mtcars.csv']))

print(cars)


              car_names   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  \
0             Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1   
1         Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1   
2            Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1   
3        Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0   
4     Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0   
5               Valiant  18.1    6  225.0  105  2.76  3.460  20.22   1   0   
6            Duster 360  14.3    8  360.0  245  3.21  3.570  15.84   0   0   
7             Merc 240D  24.4    4  146.7   62  3.69  3.190  20.00   1   0   
8              Merc 230  22.8    4  140.8   95  3.92  3.150  22.90   1   0   
9              Merc 280  19.2    6  167.6  123  3.92  3.440  18.30   1   0   
10            Merc 280C  17.8    6  167.6  123  3.92  3.440  18.90   1   0   
11           Merc 450SE  16.4    8  275.8  180  3.07  4.070  17.


car_names	mpg	cyl	disp	hp	drat	wt	qsec	vs	am	gear	carb
0	Mazda RX4	21.0	6	160.0	110	3.90	2.620	16.46	0	1	4	4
1	Mazda RX4 Wag	21.0	6	160.0	110	3.90	2.875	17.02	0	1	4	4
2	Datsun 710	22.8	4	108.0	93	3.85	2.320	18.61	1	1	4	1
3	Hornet 4 Drive	21.4	6	258.0	110	3.08	3.215	19.44	1	0	3	1
4	Hornet Sportabout	18.7	8	360.0	175	3.15	3.440	17.02	0	0	3	2

In [ ]:
#  To group a  DataFrame by its values in a particular column, call the .groupby() method off of the DataFrame, and then pass
# in the column Series you want the DataFrame to be grouped by.
cars_groups = cars.groupby(cars['cyl'])
cars_groups.mean()

<ipython-input-56-0a16fe62642b>:4: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  cars_groups.mean()


mpg        disp          hp      drat        wt       qsec  \
cyl                                                                     
4    26.663636  105.136364   82.636364  4.070909  2.285727  19.137273   
6    19.742857  183.314286  122.285714  3.585714  3.117143  17.977143   
8    15.100000  353.100000  209.214286  3.229286  3.999214  16.772143   

           vs        am      gear      carb  
cyl                                          
4    0.909091  0.727273  4.090909  1.545455  
6    0.571429  0.428571  3.857143  3.428571  
8    0.000000  0.142857  3.285714  3.500000


mpg	disp	hp	drat	wt	qsec	vs	am	gear	carb
cyl
4	26.663636	105.136364	82.636364	4.070909	2.285727	19.137273	0.909091	0.727273	4.090909	1.545455
6	19.742857	183.314286	122.285714	3.585714	3.117143	17.977143	0.571429	0.428571	3.857143	3.428571
8	15.100000	353.100000	209.214286	3.229286	3.999214	16.772143	0.000000	0.142857	3.285714	3.500000

##### chapter Quiz
1. Using the mtcars dataset, group the data by the transmission type "am". Which transmission type has the highest mean miles/(US) gallon "mpg" and its corresponding value?

cars_groups = cars.groupby(cars['am'])
cars_groups.mean()

2. What is the code used to create a subset of the mtcars dataset by indexing the columns mpg, disp, hp, and wt?

cars_subset=cars[['mpg', 'disp', 'hp', 'wt']]

3. What is the output of data.groupby(data['item_type']) data.sum()?

What is the output of data.groupby(data['item_type']) data.sum()?

4. Given a 5x5 dataframe, x_NaN, what is the code used to count the number of missing values per column and then subsequently drop those columns containing missing values?

x_NaN.isnull().sum() x_NaN = x_NaN.dropna(axis=1)

5. What is the output after concatenating dataframes, df_1 and df_2?
df_1 = np.array([[1.,3.,4.],[2.,5.,2.],[0.,4.,1.]])' and 'df_2 =np. array([[2.,1.],[0.,5.],[4.,7.]])

array([[1.,3.,4.],[2.,5.,2.],[0.,4.,1.],[2.,1.,NaN.],[0.,5.,NaN.],[4.,7.,NaN.]])

6. How would you print all the individuals in the dataframe who are younger than age 25 given a dataframe, age_obj, containing the ages of 11 individuals?

age_obj[age_obj < 25]

7. Using the data frame, df_random, what is the code used to return true/false values for all records that are less than 0.4?

df_random<0.4

This code outputs a dataframe of boolean values either true or false for all records to indicate how each element compares to the scalar value.

8. How do you drop duplicates in column 4 of the 5x5 dataframe, x_obj, and save the new dataframe as x_no_dups?

x_no_dups = x_obj.drop_duplicates(['column 4'])

9. What does the isnull() function mean?

detects NaN values

The isnull() command returns a dataframe of boolean values which are true for NaN values and false otherwise.